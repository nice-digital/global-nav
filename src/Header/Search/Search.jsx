import { useCallback } from "react";
import PropTypes from "prop-types";
import SearchIcon from "@nice-digital/icons/lib/Search";

import Autocomplete from "./Autocomplete";
import SkipLink from "../SkipLink";
import { getCallbackFunction } from "../../utils";

import styles from "./Search.module.scss";

const searchInputSelector = "header form[role='search'] [name='q']";

const Search = function ({
	url = "/search",
	autocomplete,
	skipLinkId,
	placeholder = "Search NICEâ€¦",
	query = "",
	onSearching,
	onNavigating,
}) {
	const keyDownHandler = function (e) {
		// Submit the form when we press enter to allow the enter key functionality:
		// The accessible-autocomplete doesn't let you use enter by default
		if (e.key === "Enter" && e.target.id === "autocomplete") {
			const wasSearchActionOverridden = searchSubmitHandler(e);

			if (!wasSearchActionOverridden) {
				const searchForm = document.getElementById("global-nav-search-form");
				if (searchForm) searchForm.submit();
			}
		}
	};

	/**
	 * Event handler for the form submit action
	 *
	 * @param {*} e The event argument
	 * @returns true if the search action was overridden; otherwise false
	 * @memberof Search
	 */
	const searchSubmitHandler = function (e) {
		// If we've got an onSearching prop then cancel the default search behaviour
		// and call the onSearcing callback with the query value.

		if (onSearching) {
			const onSearchingCallback = getCallbackFunction(onSearching);

			if (onSearchingCallback) {
				e.preventDefault();
				const query = document.getElementById("autocomplete").value;
				onSearchingCallback({ query });
				return true;
			}
		}

		return false;
	};

	// TODO: Remove this hack to fix https://github.com/alphagov/accessible-autocomplete/issues/434
	// We do this to make our axe tests pass
	// Wait for the search box to appear before removing the aria-activedescendant attribute
	const searchFormRef = useCallback((node) => {
		let searchInput;
		if (node && "MutationObserver" in window) {
			new MutationObserver(() => {
				searchInput =
					searchInput || document.querySelector(searchInputSelector);
				if (
					searchInput &&
					searchInput.getAttribute("aria-activedescendant") === "false"
				) {
					searchInput.setAttribute("aria-activedescendant", "");
				}
			}).observe(node, {
				attributeFilter: ["aria-activedescendant"],
				attributes: true, // See https://stackoverflow.com/a/50593541/486434
				childList: true,
				subtree: true,
			});
		}
	}, []);

	return (
		// eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
		<form
			id="global-nav-search-form"
			role="search"
			action={url}
			className={styles.search}
			onSubmit={searchSubmitHandler}
			onKeyDown={keyDownHandler}
			ref={searchFormRef}
		>
			{/* eslint jsx-a11y/label-has-for: 0 */}
			<label className={styles.label} htmlFor="autocomplete">
				{placeholder}
			</label>
			<Autocomplete
				source={autocomplete?.suggestions || autocomplete}
				suggestionTemplate={autocomplete?.suggestionTemplate}
				placeholder={placeholder}
				query={query}
				onNavigating={onNavigating}
			/>
			<button
				className={styles.button}
				type="submit"
				aria-label="Perform search"
				onClick={searchSubmitHandler}
			>
				<SearchIcon className={styles.icon} />
			</button>
			<SkipLink to={`#${skipLinkId}`}>Skip to content</SkipLink>
		</form>
	);
};

export default Search;

Search.propTypes = {
	url: PropTypes.string,
	autocomplete: PropTypes.oneOfType([
		PropTypes.shape({
			suggestions: PropTypes.oneOfType([
				PropTypes.arrayOf(
					PropTypes.shape({
						Title: PropTypes.string.isRequired,
						TitleHtml: PropTypes.string,
						TypeAheadType: PropTypes.string,
						Link: PropTypes.string.isRequired,
					})
				).isRequired,
				PropTypes.string,
			]),
			suggestionTemplate: PropTypes.func,
		}),
		PropTypes.bool,
		PropTypes.string,
		PropTypes.arrayOf(
			PropTypes.shape({
				Title: PropTypes.string.isRequired,
				TitleHtml: PropTypes.string,
				TypeAheadType: PropTypes.string,
				Link: PropTypes.string.isRequired,
			})
		),
	]),
	skipLinkId: PropTypes.string,
	placeholder: PropTypes.string,
	query: PropTypes.string,
	onSearching: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
	onNavigating: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
};
