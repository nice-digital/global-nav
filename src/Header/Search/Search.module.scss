@use 'global-nav-colours';
@use '@nice-digital/nds-core/scss/focus';
@use '@nice-digital/nds-core/scss/spacing';
@use '@nice-digital/nds-core/scss/utils';
@use '../header-settings';

.search {
  margin: 0;
  padding-right: utils.rem(header-settings.$search-height);
  position: relative;
}

.label {
  /* https://snook.ca/archives/html_and_css/hiding-content-for-accessibility */
  // TO DO: Replace with design system invisible mixin when available
  clip: rect(1px 1px 1px 1px); /* IE6, IE7 */
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute !important;
  width: 1px;
}

.button {
  background: global-nav-colours.$search-button-background;
  border: 0;
  color: global-nav-colours.$search-button-text;
  cursor: pointer;
  height: utils.rem(header-settings.$search-height);
  min-width: utils.rem(header-settings.$search-height);
  padding: utils.rem(0 spacing.$small);
  position: absolute;
  right: 0;
  text-align: center;
  top: 0;
  z-index: header-settings.$base-z-index + 2; // To sit above the search box so the focus outline surrounds the button

  &:focus {
    @include focus.default-focus-style;
  }

  .icon {
    font-size: 150%;
    vertical-align: middle;
  }
}
