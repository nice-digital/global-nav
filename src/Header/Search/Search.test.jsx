import React from "react";
import Search from "./Search";

import {
	render,
	createEvent,
	fireEvent,
	waitFor,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("Search", () => {
	const defaultProps = {};
	let appContainer;

	beforeEach(() => {
		appContainer = document.createElement("div");
		document.body.appendChild(appContainer);
	});

	afterEach(() => {
		document.body.removeChild(appContainer);
	});

	describe("Rendering", () => {
		it("Matches snapshot", () => {
			const { container } = render(
				<Search {...defaultProps} query="test" skipLinkId="content-start" />
			);
			expect(container).toMatchSnapshot();
		});
	});

	describe("Autocomplete", () => {
		it("Passes props to autocomplete component", () => {
			const { getByRole } = render(
					<Search
						{...defaultProps}
						autocomplete="variableName"
						placeholder="Test placeholder"
						query="diabetes"
					/>
				),
				input = getByRole("combobox");

			expect(input).toHaveAttribute("placeholder", "Test placeholder");
			expect(input).toHaveValue("diabetes");
		});
	});

	describe("onSearching", () => {
		describe("Form submit and button click", () => {
			it("should call the onSearching prop when the form is submitted", () => {
				const onSearching = jest.fn(),
					{ getByRole } = render(
						<Search {...defaultProps} onSearching={onSearching} />
					),
					search = getByRole("search");

				fireEvent.submit(search);

				expect(onSearching).toHaveBeenCalled();
				expect(onSearching).toHaveBeenCalledTimes(1);
			});

			it("should call onSearching prop when the button is clicked", async () => {
				const onSearching = jest.fn(),
					{ getByRole } = render(
						<Search {...defaultProps} onSearching={onSearching} />
					),
					searchButton = getByRole("button", { name: "Perform search" }),
					user = userEvent.setup();

				await user.click(searchButton);

				expect(onSearching).toHaveBeenCalled();
				expect(onSearching).toHaveBeenCalledTimes(1);
			});
		});

		it("should prevent default and call callback with the search query for a onSearching function", () => {
			const onSearching = jest.fn(),
				{ getByRole } = render(
					<Search {...defaultProps} query="test" onSearching={onSearching} />
				),
				search = getByRole("search"),
				submitEvent = createEvent.submit(search);

			fireEvent(search, submitEvent);

			expect(submitEvent.defaultPrevented).toBe(true);
			expect(onSearching).toHaveBeenCalledWith({ query: "test" });
		});

		it("should prevent default and call callback for a named onSearching global function", () => {
			const onSearching = jest.fn();
			window.onSearchingHandler = onSearching;

			const { getByRole } = render(
					<Search
						{...defaultProps}
						query="test"
						onSearching="onSearchingHandler"
					/>
				),
				search = getByRole("search"),
				submitEvent = createEvent.submit(search);

			fireEvent(search, submitEvent);

			expect(submitEvent.defaultPrevented).toBe(true);
			expect(onSearching).toHaveBeenCalledWith({ query: "test" });
		});

		it("should not prevent default with no onSearching prop", () => {
			const { getByRole } = render(<Search {...defaultProps} />),
				search = getByRole("search"),
				submitEvent = createEvent.submit(search);

			fireEvent(search, submitEvent);

			expect(submitEvent.defaultPrevented).toBe(false);
		});

		it("should not prevent default with missing onSearching global function prop", () => {
			const { getByRole } = render(
					<Search {...defaultProps} onSearching="blahblah" />
				),
				search = getByRole("search"),
				submitEvent = createEvent.submit(search);

			fireEvent(search, submitEvent);

			expect(submitEvent.defaultPrevented).toBe(false);
		});
	});

	describe("enter key", () => {
		it("should call onSearching callback prop if the enter key is pressed", async () => {
			const onSearching = jest.fn(),
				{ getByRole } = render(
					<Search {...defaultProps} onSearching={onSearching} />
				),
				input = getByRole("searchbox"),
				user = userEvent.setup();

			await user.type(input, "{enter}");

			expect(onSearching).toHaveBeenCalledTimes(1);
		});
	});
});
