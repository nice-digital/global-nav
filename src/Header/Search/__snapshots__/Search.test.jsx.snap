// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Rendering Matches snapshot 1`] = `
<div>
  <form
    action="/search"
    class="search"
    id="global-nav-search-form"
    role="search"
  >
    <label
      class="label"
      for="autocomplete"
    >
      Search NICE…
    </label>
    <div
      class="ac"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        style="position: absolute; left: -9999px; height: 1px; width: 1px; overflow: hidden;"
      />
      <div
        class="autocomplete__wrapper"
      >
        <input
          class="autocomplete__input autocomplete__input--default"
          data-hj-allow=""
          id="autocomplete"
          maxlength="512"
          name="q"
          placeholder="Search NICE…"
          type="search"
          value="test"
        />
      </div>
    </div>
    <button
      aria-label="Perform search"
      class="button"
      type="submit"
    >
      <svg
        aria-hidden="true"
        class="icon"
        fill="none"
        focusable="false"
        height="1em"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M83.784 476.472 197.112 363.16c30.144 20.32 65.056 30.528 100.032 30.528 45.904 0 91.776-17.488 126.8-52.528 70.064-70.064 70.064-183.584 0-253.616C388.936 52.504 343.048 35 297.144 35c-45.904 0-91.808 17.504-126.848 52.528-61.696 61.696-68.976 157.12-21.968 226.864L35 427.752l48.784 48.72Zm213.36-137.952c-33.168 0-64.352-12.896-87.792-36.368-48.416-48.448-48.416-127.2 0-175.616 23.44-23.44 54.64-36.352 87.792-36.352s64.336 12.912 87.776 36.368c23.44 23.456 36.384 54.64 36.384 87.792s-12.928 64.352-36.384 87.792c-23.456 23.472-54.624 36.368-87.776 36.368v.016Z"
          fill="currentColor"
        />
      </svg>
    </button>
    <a
      class="link"
      href="#content-start"
    >
      Skip to content
    </a>
  </form>
</div>
`;
